### PR TITLE
Replace icu_data_key macro with const DataKeys

### LINKED ISSUE #212: Data provider: match on TinyStr instead of &str
We can be more efficient by matching on TinyStr by integer value rather than &str.

Blocked on https://github.com/zbraniecki/tinystr/issues/17

CC @zbraniecki

### PR DESCRIPTION
**Pull request by sffc**
Fixes #212

### CODE REVIEW

File: components/provider/src/data_key.rs [hunk 1]
```diff
@@ -185,29 +147,30 @@ impl From<&DataKey> for DataKeyComponents {
...
         },
         TestCase {
-            data_key: icu_data_key!(x-private: cardinal@1),
-            expected: "x-private/cardinal@1",
-        },
-        TestCase {
-            data_key: icu_data_key!(x-maxlengthprivate: cardinal@1),
-            expected: "x-maxlengthprivate/cardinal@1",
+            data_key: DataKey {
+                category: DataCategory::PrivateUse(tinystr4!("priv")),
```
Comment (zbraniecki) [inline components/provider/src/data_key.rs:162]: can you add a factory method to generate it out of str instead?

**Review by zbraniecki — APPROVED** [review]

File: components/provider/src/data_key.rs (hunk 1 above)
Comment (sffc) [reply components/provider/src/data_key.rs:162]: I could, but private use is not a common use case (we don't want people accidentally converting a string to private use), and you would lose the constness that we get by spelling it out with the macro.

Comment (zbraniecki) [reply components/provider/src/data_key.rs:162]: Yeah, I wouldn't do it as a default, but something like `DataCategory::newPrivateUse(&str)` or something like that. It's optional, and for later, just that using TinyStr in public API is a biit iffy for me.

**Review by zbraniecki — APPROVED** [review]

File: components/provider/src/data_key.rs (hunk 1 above)
Comment (sffc) [reply components/provider/src/data_key.rs:162]: 1. It's an enum, so the TinyStr field is public
2. I don't see anything wrong with having TinyStr be public
3. `DataCategory::newPrivateUse(&str)` can't be const, but the common use for DataKey and DataCategory is that they are const

Comment (zbraniecki) [reply components/provider/src/data_key.rs:162]: sgtm

**Pull request merged**